test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/TransportLayerTests/RequestMessageArgsTests.cs

```diff
@@ -0,0 +1,121 @@
+
+                services.AddControllers().AddOData(opt => opt.Count().Filter().Expand().Select().OrderBy().SetMaxTop(null)
+                    .AddRouteComponents("odata", CommonEndToEndEdmModel.GetEdmModel()));
+            }
+        }
+
+        public RequestMessageArgsTests(TestWebApplicationFactory<TestsStartup> fixture)
+            : base(fixture)
+        {
+            _baseUri = new Uri(Client.BaseAddress, "odata/");
+
+            _context = new Container(_baseUri)
+            {
+                HttpClientFactory = HttpClientFactory
+            };
+
+            ResetDataSource();
+        }
+
+        [Fact]
+        public async Task AddingCustomHeaderToRequest_ShouldSucceed()
+        {
+            string headerName = "MyNewHeader";
+            string headerValue = "MyNewHeaderValue";
+
+            Func<DataServiceClientRequestMessageArgs, DataServiceClientRequestMessage> configureRequest =
+                args =>
+                {
+                    args.Headers.Add(headerName, headerValue);
+                    return new HttpClientRequestMessage(args);
+                };
+
+            this.UpdateContext(configureRequest);
+
+            DataServiceQuery<ClientEndToEndModel.Product> query = _context.Products;
+            await query.ExecuteAsync();
+
+            Assert.NotNull(this.lastRequestMessage);
+            var header = this.lastRequestMessage.Headers.SingleOrDefault(h => h.Key == headerName);
+            Assert.Equal(headerValue, header.Value);
+        }
+
+        [Fact]
+        public async Task ChangingMergeToPatch_ShouldUpdateRequestMethod()
+        {
+            Func<DataServiceClientRequestMessageArgs, DataServiceClientRequestMessage> configureRequest =
+                args =>
+                {
+                    if (args.Method == "MERGE")
+                    {
+                        var newArgs = new DataServiceClientRequestMessageArgs("PATCH", args.RequestUri, true, args.Headers);
+
+                        // Ensure compatibility with V3
+                        newArgs.Headers.Remove("DataServiceVersion");
+                        newArgs.Headers.Add("DataServiceVersion", "3.0");
+
+                        return new HttpClientRequestMessage(newArgs);
+                    }
+
+                    return new HttpClientRequestMessage(args);
+                };
+
+            this.UpdateContext(configureRequest);
+
+            var query = _context.Products.Take(1) as DataServiceQuery<ClientEndToEndModel.Product>;
+            var product = (await query.ExecuteAsync()).Single();
+
+            this.lastRequestMessage = null;
+            product.Description = "New Description " + Guid.NewGuid();
+            _context.UpdateObject(product);
+            await _context.SaveChangesAsync();
+
+            Assert.NotNull(this.lastRequestMessage);
+            Assert.Equal("PATCH", this.lastRequestMessage.Method);

```

Comment:
How do we know if the custom config was called? And how come we're not checking if the new header you added was actually added?

Reply:
I think the custom config will never be called since this is a v3 construct. I don't think this test is needed here.